indexer/centers_table.cpp

```diff
@@ -0,0 +1,337 @@
+{
+  WriteToSink(writer, m_version);
+  WriteToSink(writer, m_endiannes);
+}
+
+bool CentersTable::Header::IsValid() const
+{
+  if (m_endiannes > 1)
+    return false;
+  return true;
+}
+
+// CentersTable ------------------------------------------------------------------------------------
+unique_ptr<CentersTable> CentersTable::Load(Reader & reader,
+                                            serial::CodingParams const & codingParams)
+{
+  uint16_t const version = ReadPrimitiveFromPos<uint16_t>(reader, 0 /* pos */);
+  if (version != 0)
+    return unique_ptr<CentersTable>();
+
+  // Only single version of centers table is supported now.  If you
+  // need to implement new versions of CentersTable, implement
+  // dispatching based on first-four-bytes version.
+  unique_ptr<CentersTable> table = make_unique<CentersTableV0>(reader, codingParams);
+  if (!table->Init())
+    return unique_ptr<CentersTable>();
+  return table;
+}
+
+// CentersTableBuilder -----------------------------------------------------------------------------
+CentersTableBuilder::CentersTableBuilder(Writer & writer, serial::CodingParams const & codingParams)
+  : m_writer(writer), m_codingParams(codingParams)
+{
+}
+
+CentersTableBuilder::~CentersTableBuilder()
+{
+  CentersTableV0::Header header;
+
+  int64_t const startOffset = m_writer.Pos();
+  header.Write(m_writer);
+
+  {
+    uint64_t const numBits = m_ids.empty() ? 0 : m_ids.back() + 1;
+
+    succinct::bit_vector_builder builder(numBits);
+    for (auto const & id : m_ids)
+      builder.set(id, true);
+
+    coding::FreezeVisitor<Writer> visitor(m_writer);
+    succinct::rs_bit_vector(&builder).map(visitor);
+  }
+
+  vector<uint32_t> offsets;
+  vector<uint8_t> deltas;
+
+  {
+    MemWriter<vector<uint8_t>> writer(deltas);
+    for (size_t i = 0; i < m_centers.size(); i += CentersTableV0::kFrequency)
+    {
+      offsets.push_back(deltas.size());
+
+      uint64_t delta = EncodeDelta(m_centers[i], m_codingParams.GetBasePoint());
+      WriteVarUint(writer, delta);
+      for (size_t j = i + 1; j < i + CentersTableV0::kFrequency && j < m_centers.size(); ++j)
+      {
+        delta = EncodeDelta(m_centers[j], m_centers[j - 1]);
+        WriteVarUint(writer, delta);
+      }
+    }
+  }
+
+  {
+    succinct::elias_fano::elias_fano_builder builder(offsets.empty() ? 0 : offsets.back() + 1,
+                                                     offsets.size());
+    for (auto const & offset : offsets)
+      builder.push_back(offset);
+
+    header.m_positionsOffset = m_writer.Pos();
+    coding::FreezeVisitor<Writer> visitor(m_writer);
+    succinct::elias_fano(&builder).map(visitor);
+  }
+
+  {
+    header.m_deltasOffset = m_writer.Pos();
+    m_writer.Write(deltas.data(), deltas.size());
+    header.m_endOffset = m_writer.Pos();
+  }
+
+  int64_t const endOffset = m_writer.Pos();
+
+  m_writer.Seek(startOffset);
+  m_writer.Write(&header, sizeof(header));
+  m_writer.Seek(endOffset);
```

Comment:
Yes, it's worth. There're no guarantees that the caller will not use writer after the call.